File: 0198-house-robber/solution.cpp
class Solution {
public:
    int dp[101];
    int rob(vector<int>& nums) {
        memset(dp , -1 , sizeof dp);
        const int n = nums.size() ;
        return go(nums , n - 1);
    }
    
    int go(vector<int> v , int idx){
        
        if(idx < 0 )return 0 ;
       // if(idx < 1 ) return v[0] ;
        
        if( dp[idx] != -1) return dp[idx];
        
        int x = go(v, idx - 2) + v[idx] ;
        int y = go(v,idx -1 ) ;
        
        return dp[idx] = max(x,y);
        
    }
    
};

File: 0198-house-robber/solution.java
class Solution {
    int[] dp ;
    public int rob(int[] nums) {
        dp = new int[101];
        for(int i = 0 ; i < 101 ; i++) dp[i] = -1 ;
        return solve(nums,0);
    }

    public int solve(int[] nums , int idx){
        if(idx >= nums.length) return 0;
        if(dp[idx] != -1) return dp[idx];
        return dp[idx] = Math.max( nums[idx] + solve(nums,idx + 2) , solve(nums,idx + 1));
    }
}
